Userland/Libraries/LibWeb/HTML/Navigation.cpp

```diff
@@ -137,6 +174,97 @@ bool Navigation::can_go_forward() const
+    auto state = options.state.value_or(JS::js_undefined());
+
+    // 5. Let serializedState be StructuredSerializeForStorage(state).
+    //    If this throws an exception, then return an early error result for that exception.
+    // FIXME: Fix this spec grammaro in the note
+    // NOTE: It is importantly to perform this step early, since serialization can invoke web developer code,
+    //       which in turn might change various things we check in later steps.
+    auto serialized_state_or_error = structured_serialize_for_storage(vm, state);
+    if (serialized_state_or_error.is_error()) {
+        return early_error_result(serialized_state_or_error.release_error());
+    }
+
+    auto serialized_state = serialized_state_or_error.release_value();
+
+    // 6. If document is not fully active, then return an early error result for an "InvalidStateError" DOMException.
+    if (!document.is_fully_active())
+        return early_error_result(WebIDL::InvalidStateError::create(realm, "Document is not fully active"));
+
+    // 7. If document's unload counter is greater than 0, then return an early error result for an "InvalidStateError" DOMException.
+    if (document.unload_counter() > 0)
+        return early_error_result(WebIDL::InvalidStateError::create(realm, "Document already unloaded"));
+
+    // 8. Let info be options["info"], if it exists; otherwise, undefined.
+    auto info = options.info.value_or(JS::js_undefined());
+
+    // 9. Let apiMethodTracker be the result of maybe setting the upcoming non-traverse API method tracker for this
+    //    given info and serializedState.
+    auto api_method_tracker = maybe_set_the_upcoming_non_traverse_api_method_tracker(info, serialized_state);
+
+    // 10. Navigate document's node navigable to urlRecord using document,
+    //     with historyHandling set to options["history"] and navigationAPIState set to serializedState.
+    // FIXME: Fix spec typo here
+    // NOTE: Unlike location.assign() and friends, which are exposed across origin-domain boundaries,
+    //       navigation.navigate() can only be accessed by code with direct synchronous access to the
+    ///      window.navigation property. Thus, we avoid the complications about attributing the source document
+    //       of the navigation, and we don't need to deal with the allowed by sandboxing to navigate check and its
+    //       acccompanying exceptionsEnabled flag. We just treat all navigations as if they come from the Document
+    //       corresponding to this Navigation object itself (i.e., document).
+    auto history_handling_behavior = to_history_handling_behavior(options.history);
+    // FIXME: This signature is wrong. We need to pass in the serialized state.
+    //        ... Can we create an overload with just these properties? (url, document, history behavior, state)
+    MUST(document.navigable()->navigate(url_record, document, {}, nullptr, false, history_handling_behavior));
```

Comment:
Should I just make this real dealio fixme until we have Navigables hooked up such that the spec assumption that we actually have one on this document is true? That would cause this method to always return two rejected progress with the AbortError below

Reply:
yes, it looks ok for me.